### ISSUE TITLE
Advanced load balancing

### ISSUE THREAD

**Issue by ankoh** [istio/istio#11813]
We are currently investigating, how we can integrate replicated database servers in an Istio service mesh.
I'm convinced that envoy itself is a perfect fit for us but the interplay with Istio is holding us back a bit.
We would use envoy primarily as L4 load balancer with a custom network filter to inspect the database protocol.
From skimming the docs, I got the strong impression that Istio is primarily built around the HTTP features of envoy.
Istio probably [lets us control the configuration of our network filter](https://istio.io/docs/reference/config/istio.networking.v1alpha3/#EnvoyFilter) but I assume that the load balancing will become a problem.
Long-running TCP connections will rather be the norm than an exception and might induce varying load on the database pods.
So the load balancer MUST consider the load on these pods and CANNOT just route incoming connections with the current rather [simple load balancing strategies](https://istio.io/docs/reference/config/istio.networking.v1alpha3/#LoadBalancerSettings).

With my still quite incomplete understanding of what Istio can do, I thought about 3 different approaches:
1) The envoy EDS [endpoint.LbEndpoint](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/endpoint/endpoint.proto#endpoint-lbendpoint) supports a `load_balancing_weight`. We could consume the Mixer telemetry and adjust these `load_balancing_weight`s but if I am correct Istio currently only supports weighted load balancing between localities?
2) We could somehow try to access the Mixer telemetry in our envoy filters directly and do the load balancing manually.
3) We could use a side channel to get the load information to the envoy proxies (if mixer won't work).

Did I miss something here?
I'm very much leaning towards 1) but this option rises and falls with the feasibility to bring this load balancing to pilot/mixer.
Do you have plans in this area?
Envoy recently started to work on [MySQL](https://www.google.com/search?client=ubuntu&channel=fs&q=envoy+mysql&ie=utf-8&oe=utf-8) & [Postgres](https://github.com/envoyproxy/envoy/issues/2861) network filters which will lead to this exact same problem as well.

**Comment by rshriram** [comment]
if you represent the database as a service entry with endpoints for each peer in the pool, you can add weight for each endpoint in the service entry's endpoint. You can dynamically adjust these weights based on your telemetry info and update the service entry spec. A standard weighted round robin https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing/load_balancers should satisfy your load based endpoint selection [you might have to invert the weights]

You might have to fix the CDS/EDS code to use the weight in the endpoint when shipping EDS/CDS for all clusters.

**Comment by ankoh** [comment]
Interesting, thanks.
But that would force me to do the endpoint discovery on my own right?
This is quite undesirable especially as this cuts even more from Istio's feature set.

**Comment by ankoh** [comment]
I just learned that Envoy supports HTTP/2 CONNECT.
https://github.com/envoyproxy/envoy/issues/1630

With that, this feature request boils down to something like **intra cluster load balancing with mixer telemetry**.
If you think this is feasible without too much architectural rework and want to give me some first hints, I'd be happy to contribute some code.

**Comment by ankoh** [comment]
On the other hand, deriving endpoint weights should probably not be generalized as there are too many moving parts involved.
I'll probably just do the endpoint discovery and weight adjustment externally for now.

Thanks.

**Comment by ankoh** [comment]
I just saw that Envoy is getting support for weighted consistent hashing very soon.
https://github.com/envoyproxy/envoy/pull/5846/files

It would be perfect for us, if there was a way to adjust the endpoint weights of the cluster load assignment in pilot via an API.

**Comment by rshriram** [comment]
You dont have to do endpoint discovery if you use stateful sets. All members of a stateful set get a stable DNS name (peer-1.mystatefulset.ns1.svc.cluster.local). So your service entry could be like this:

```
serviceEntry:
host: *.mystatefulset.ns1.svc.cluster.local
port: 3306/mysql
resolution: DNS
endpoints:
- peer-1.mystatefulset.ns1.svc.cluster.local
   weight: 80 #adjust this dynamically
- peer-2.mystatefulset.ns1.svc.cluster.local
   weight: 20
---
Sidecar to create a special outbound listener for 3306:
egress:
- port: 3306
   hosts:
   - "*.mystatefulset.n1.svc.cluster.local"
- hosts:
   - */* # catch all
---
```

**Comment by ankoh** [comment]
@rshriram I'm not quite sure whether I like that Idea. Especially in combination with autoscaling, I would need a Kubernetes controller that adjusts the endpoint weights as soon as new endpoints are joining the set right?
The alternative would be to proxy XDS ADS endpoint between envoy and pilot and inject the endpoint weights there.
Neither solution is really beautiful to be honest.

Is it correct that Pilot itself is meant to be stateless?
So that non-XDS Pilot API would rather contradict the design philosophy behind pilot as the state would no longer live solely in Kubernetes?

**Comment by ankoh** [comment]
I picked up an Idea that floated around in the Envoy issues (also mentioned by you I think) to just tunnel the database connection over HTTP/2.
I did that already and it works quite beautifully.
It allows us to do consistent hashing on arbitrary connection parameters, which are just wrapped in a single additional http header per connection.

But even more interesting is this recent PR, which would allow us to do more fine grained load balancing by using more reasonable endpoint weights.
https://github.com/envoyproxy/envoy/pull/5846

If there only was a way to inject these endpoint weights properly.
With my current understanding how Pilot works right now, I'm not quite sure whether one should try to bring this functionality to Pilot.

Which leaves me with 2 options, I think: Either proxying the XDS API or the manual service entry that you proposed earlier.

**Comment by ankoh** [comment]
I'll close this here. It turned out to be far easier to just implement the routing with a custom tcp_proxy.
I liked the idea of tunnelling via HTTP/2 and rely on the HTTP router but I found it quite tedious to hijack the WebSocket upgrade support for our needs.

A few things that I noticed:
- HTTP2 tunnelling could be quite useful for protocols that are not HTTP1.1 compliant.
- Manually "downgrading" the WebSocket is still quite annoying as we cannot just plug in Network filters in the http_connection_manager.
- This is probably due to the fact that websockets just feel a little bit weird in the Request/Response-setting of the StreamDecoders/Encoders.
- The current route discovery is very tailored towards HTTP but tunnelling arbitrary (non HTTP1.1 protocols) over HTTP2 would probably benefit a lot from a more abstract router. People could just extract the necessary metadata from their protocols with a simple network filter and provide it to this router instead of stiffing everything into a synthetic HTTP header.
